icu4c/source/common/unicode/ulocbuilder.h

```diff
@@ -0,0 +1,401 @@
+ * @param length the length of the variant; if negative, then the variant need to be
+ *               null terminated,
+ * @draft ICU 74
+ */
+U_CAPI void U_EXPORT2
+ulocbld_setVariant(ULocaleBuilder* builder, const char* variant, int32_t length);
+
+/**
+ * Sets the extension for the given key. If the value is the empty string,
+ * the extension is removed.  Otherwise, the <code>key</code> and
+ * <code>value</code> must be well-formed, or else the ulocbld_buildLocaleID()
+ * and ulocbld_buildLanguageTag() methods will
+ * later report an U_ILLEGAL_ARGUMENT_ERROR.
+ *
+ * <p><b>Note:</b> The key ('u') is used for the Unicode locale extension.
+ * Setting a value for this key replaces any existing Unicode locale key/type
+ * pairs with those defined in the extension.
+ *
+ * <p><b>Note:</b> The key ('x') is used for the private use code. To be
+ * well-formed, the value for this key needs only to have subtags of one to
+ * eight alphanumeric characters, not two to eight as in the general case.
+ *
+ * @param builder the builder
+ * @param key the extension key
+ * @param value the value, a const char * pointer (need not be terminated when
+ *               the length is non-negative)
+ * @param length the length of the value; if negative, then the value need to be
+ *               null terminated,
+ * @draft ICU 74
+ */
+U_CAPI void U_EXPORT2
+ulocbld_setExtension(ULocaleBuilder* builder, char key, const char* value, int32_t length);
+
+/**
+ * Sets the Unicode locale keyword type for the given key. If the type
+ * StringPiece is constructed with a nullptr, the keyword is removed.
+ * If the type is the empty string, the keyword is set without type subtags.
+ * Otherwise, the key and type must be well-formed, or else the
+ * ulocbld_buildLocaleID() and ulocbld_buildLanguageTag() methods will later
+ * report an U_ILLEGAL_ARGUMENT_ERROR.
+ *
+ * <p>Keys and types are converted to lower case.
+ *
+ * <p><b>Note</b>:Setting the 'u' extension via {@link #setExtension}
+ * replaces all Unicode locale keywords with those defined in the
+ * extension.
+ *
+ * @param builder the builder
+ * @param key the Unicode locale key, a const char * pointer (need not be
+ *               terminated when the length is non-negative)
+ * @param keyLength the length of the key; if negative, then the key need to be
+ *               null terminated,
+ * @param type the Unicode locale type, a const char * pointer (need not be
+ *               terminated when the length is non-negative)
+ * @param typeLength the length of the type; if negative, then the type need to
+ *               be null terminated,
+ * @return This builder.
+ * @draft ICU 74
+ */
+U_CAPI void U_EXPORT2
+ulocbld_setUnicodeLocaleKeyword(ULocaleBuilder* builder,
+        const char* key, int32_t keyLength, const char* type, int32_t typeLength);
+
+/**
+ * Adds a unicode locale attribute, if not already present, otherwise
+ * has no effect.  The attribute must not be empty string and must be
```

Comment:
Dumb question: What's a "Unicode locale attribute"?

Reply:
"attribute = alphanum{3,8} ;	" in unicode_locale_extensions
https://www.unicode.org/reports/tr35/tr35.html#unicode_locale_extensions

A 'u' extension may contain multiple attribute s or keyword s as defined in [Unicode locale identifier](https://www.unicode.org/reports/tr35/tr35.html#Unicode_locale_identifier). The canonical syntax is defined as in [Canonical Unicode Locale Identifiers](https://www.unicode.org/reports/tr35/tr35.html#Canonical_Unicode_Locale_Identifiers).

Also see 
https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/ULocale.Builder.html#addUnicodeLocaleAttribute-java.lang.String-

https://download.java.net/java/early_access/jdk21/docs//api/java.base/java/util/Locale.Builder.html#addUnicodeLocaleAttribute(java.lang.String)